Scale extruder motor current linearly with speed.

mkbel opened the pull request:
49% less heating when running at low speed and standstill, 4% more torque at maximum extrusion rate (15mm^3/s), 15% more torque in high speed movements (un/retractions).

StealthChop mode is used for low speeds (below 900mm/min)
spreadCycle is used above. Transition speed is well above maximum extrusion rate of 15mm^3/s (275mm/min) so mode transition is not expected to be visible on printed surface.

StealthChop is expected to improve printed surface quality (less artifacts).

Warning you can burn extruder motor if it is not the same impedance as original Prusa i3 Extruder stepper motor. There is no current feedback in low speed so lower impedance motor can be burned by over current.

Even there is no direct current feedback, there is no risk for original motor thermal runaway, as motor resistance increases with temperature, current decreases.

Standstill peak phase current is expected to be 500 mA and linearly increase with speed to 970 mA at 900mm/min where spreadCycle constant current regulation takes over and keeps peak current at 805 mA to maximum speed possible.

As motor heating increases with current squared, lowering low speed current from 700mA to 500mA decreases heating 49% in thate mode, where motor spends most of the time.

PFW-1234

mkbel committed:
Scale extruder motor current linearly with speed.

49% less heating when running at low speed and standstill, 4% more torque at maximum extrusion rate (15mm^3/s), 15% more torque in high speed movements (un/retractions).

StealthChop mode is used for low speeds (below 900mm/min)
spreadCycle is used above. Transition speed is well above maximum extrusion rate of 15mm^3/s (275mm/min) so mode transition is not expected to be visible on printed surface.

StealthChop is expected to improve printed surface quality (less artifacts).

Warning you can burn extruder motor if it is not the same impedance as original Prusa i3 Extruder stepper motor. There is no current feedback in low speed so lower impedance motor can be burned by over current.

Even there is no direct current feedback, there is no risk for original motor thermal runaway, as motor resistance increases with temperature, current decreases.

Standstill peak phase current is expected to be 500 mA and linearly increase with speed to 970 mA at 900mm/min where spreadCycle constant current regulation takes over and keeps peak current at 805 mA to maximum speed possible.

As motor heating increases with current squared, lowering low speed current from 700mA to 500mA decreases heating 49% in thate mode, where motor spends most of the time.

mkbel commented:
Test instructions: Use factory reset to be sure new increased spread cycle current is applied. (Factory reset is not needed for printers in farm mode, as spreadCycle current is now the same as in farm mode.) Do not interfere with current from gcode (Do not issue M907 in gcode.).

3d-gussner commented:
> Test instructions: Use factory reset to be sure new increased spread cycle current is applied. Do not interfere with current from gcode (Do not issue M907 in gcode.).

As Prusa Slicer settings [1.1.8](https://github.com/prusa3d/PrusaSlicer-settings/commit/dc0a93992b69bd05a0741a6a90ba3e84855686ae) added to start and end gcode `M907` commands, you should
create a modifiled Printer profile to test it without
Start g-code
```
; Don't change E values below. Excessive value can damage the printer.
{if print_settings_id=~/(0.05mm ULTRADETAIL @MK3|0.07mm ULTRADETAIL @MK3|0.10mm DETAIL @MK3|0.15mm QUALITY @MK3|0.20mm QUALITY @MK3)/}M907 E430 ; set extruder motor current{endif}
{if print_settings_id=~/(0.15mm SPEED @MK3|0.20mm SPEED @MK3|0.30mm DRAFT @MK3)/}M907 E538 ; set extruder motor current{endif}
```
End g-code
```
{if print_settings_id=~/.*(DETAIL @MK3|QUALITY @MK3|@0.25 nozzle MK3).*/}M907 E538 ; reset extruder motor current{endif}
```
Or comment out these lines with `;`

@mkbel correct?

mkbel commented:
> @mkbel correct?

Correct. That is why I mentioned it.

mkbel commented:
Extruder motor one phase current - printing long straight segments:
(left side faster extrusion, higher peak current) (right side slower extrusion, lower peak current)
![SCR04](https://user-images.githubusercontent.com/35807926/91179219-82d3b300-e6e6-11ea-9111-889df1949539.PNG)
Extruder motor one phase current - printing short segments:
![SCR03](https://user-images.githubusercontent.com/35807926/91179394-b7e00580-e6e6-11ea-8b99-f998dac3e203.PNG)

mkbel commented:
@wavexx asked: Do you actually see/hear a jerk during the transition?
Transition is relatively smooth.
Current waveform when switching between stealthChop and spreadCycle:
(transition visible in first half-wave) G1 F6400 E-10
![WA000001](https://user-images.githubusercontent.com/35807926/91190497-edd7b680-e6f3-11ea-815f-cf0653cd9afb.png)
(transition visible in last half-wave) G1 F6400 E-10
![WA000002](https://user-images.githubusercontent.com/35807926/91190501-efa17a00-e6f3-11ea-9609-7a33387e8725.png)
(transition visible in first and last half-wave) G1 F6400 E-7
![WA000003](https://user-images.githubusercontent.com/35807926/91190511-f16b3d80-e6f3-11ea-9adc-98bc9c6e58bf.png)

mkbel commented:
@wavexx asked: When you set the transition speed between stealthchop and spreadcycle, did you try to do it at lower speeds? No I didn't try any other transition speed. I spend several hours tuning currents and I was very happy with first transition speed I tried. There are several constrains on transition speed. I don't want transition speed to be in speed used for extruding to avoid artifacts, so it must be higher than 400 mm/min. It must be also lower than 2200 mm/min as in that speed Marlin starts issuing two micro steps at once and it totally breaks voltage PWM regulation and motor phase current skyrockets to 1800 mA (nearly twice as rated current). Another constrain is, that spread cycle regulates constant phase current, so transition should be in area where voltage PWM gets similar current to have smooth transition.

mkbel commented:
@3d-gussner asked: How did you get the temperature and torque values? How looks your test rig?
I measured phase current with scope and current probe. Temperature rise is expected to be proportional to current squared. Torque is expected to be proportional to current.

mkbel commented:
@3d-gussner asked: Did you account the tolerances of the e motor? There are +-10% for resistance and +-20 for inductance in the data sheet.
Those effects are hard to predict, there must be some large scale test on our farm. I think it would be enough to check actual motor heating and if it decreased from previous firmware. And of course print quality. I don't worry that some current spikes - if present - could damage the motor. Short term over current is not damaging for the wire of motor coils. What is limiting for motor operation is maximum running temperature. We are currently very far from motor maximum operating temperature of 130°C and we hopefully will be even farther from it if this change is successful.
If extruder motor runs very slowly - which is what it does most of the time major effect on motor current has its resistance and it is also mostly the resistance which causes heating. In previous state there was constant current regulation, so current was expected to be constant and its variation between printers was affected only by accuracy of feedback resistor and reference voltage. If we assert that this variance was much smaller than motor resistance variance different users got  E motor increase in temperature which varies +-10% between printers before and constant torque. After the change, there is constant voltage, so dependence between heating power loses and resistance is inverted. Extruder motors with 10%+ resistance which heated most before will heat 10% less than average, but also its torque will be decreased 10% below average. Extruder motors with 10%- resistance which heated least before will heat 10% more than average (or even more if current gets into area where its steel becomes magnetically saturated), but also its torque will be increased 10% above average (if not saturated).

leptun commented:
Important note: If you change the steps/mm of the Extruder or change the microstepping of the motor you will have to adjust the TMC2130_TPWMTHRS_E value.

mkbel commented:
> Important note: If you change the microstepping of the motor you will have to adjust the TMC2130_TPWMTHRS_E value.

I am not sure about that, as TPWMTHRS compares to TSTEP and it is supposed to be "Actual  measured  time  between  two  1/256  micro-steps  derived from  the  step  input  frequency  in  units  of  1/fCLK." So it should be independent of micro-stepping settings according to datasheet.

But if you use higher micro-step resolution Marlin starts issuing two micro steps at once in lower speed and it totally breaks voltage PWM regulation and motor phase current skyrockets to 1800 mA.

mkbel commented:
> Important note: If you change the steps/mm of the Extruder motor you will have to adjust the TMC2130_TPWMTHRS_E value.

In such case you have probably different E motor and you can not use my PL at all. (Or you need to tune PWM_AMPL_E and PWM_GRAD_E yourself.)
Warning you can burn extruder motor if it is not the same impedance as original Prusa i3 Extruder stepper motor. There is no current feedback in low speed so lower impedance motor can be burned by over current.

DRracer commented:
**To prevent new conspiracy theories, which arose again on various social media...**
This PR is a purely experimental piece of code which might help lowering the E-motor's temperature. 
It is not meant to be merged in near future.

CacikeOtoao commented:
@DRacer

How fine tuned is this experimental feature in regards to impedance variance tolerance? +/- 10%? 20%?. I’m just trying to see if it is something that will have a wide adaptive curve for various motors or if it is more LDO/Prusa motor specific. Not bashing on anyone/anything, just want to prepare for it and test with various steppers.

thisiskeithb commented:
> Warning you can burn extruder motor if it is not the same impedance as original Prusa i3 Extruder stepper motor. There is no current feedback in low speed so lower impedance motor can be burned by over current.

Will there be a way to disable this feature for Bondtech BMG/LDO 0.9° pancake stepper upgrades or other "non-original" motors or will it require modifying firmware?

Panayiotis-git commented:
> This PR is a purely experimental piece of code which might help lowering the E-motor's temperature.
It is not meant to be merged in near future.

mkbel commented:
> Will there be a way to disable this feature for Bondtech BMG/LDO 0.9° pancake stepper upgrades or other "non-original" motors or will it require modifying firmware?

I would suggest adding menu item for that:
Rename existing menu item "mode" to "mode XYZ".
Add menu item "mode E" with options [normal] and [adaptive].
Make "mode E" [normal] default option. (After firmware upgrade and factory reset)
Add warning dialog when switching "mode E" from [normal] to [adaptive]:
"You can burn E motor and/or driver"
"if your E motor is lower impedance"
"or back EMF than ORIGINAL PRUSA"
"i3 Extruder Stepper motor."
"Do you wish to proceed?"
" Yes >No "

I have no final word if the whole extruder motor scaling thing will ever make it to official firmware and how it will be presented.

mkbel commented:
> How fine tuned is this experimental feature in regards to impedance variance tolerance? +/- 10%? 20%?. I’m just trying to see if it is something that will have a wide adaptive curve for various motors or if it is more LDO/Prusa motor specific. Not bashing on anyone/anything, just want to prepare for it and test with various steppers.

It is fine tuned to one specimen of ORIGINAL PRUSA i3 Extruder Stepper motor. It is not yet even known, if hard coded settings will be usable across parameter dispersion across same motor model. It was tested on two other printers with same motor model and no magical smoke was released and printers were actually printing. It is highly probable that it will work  across parameter dispersion across same motor model.

It is highly improbable that it could work with different motor model with different impedance and back EMF.

The whole feature is meant as workaround for too hot E motor. I personally never encountered problem related to too hot E motor. So I expect only minority of users are affected by this problem. I made this to make happy those minority users and I don't expect any use of this feature for most of the users. I expect the majority of users unaffected with hot E motor related problems will stay with constant current regulation as before.

leptun commented:
I’ve included these changes in my Klipper config and it worked correctly for me. My hardware was an original prusa stepper (high inductance) from the Mk2s era with an EinsyRambo.
The motor was warm, but not as hot as it was on my other Mk3 with stock firmware and current settings.

mkbel commented:
It was decided to:

- put this option to experimental menu
- always enable in farm_mode
- if enabled ignore M907 E gcode

mkbel commented:
Because of conflicting requirements to be always enabled in farm_mode and not to be possible to enable via gcode, additional safety mechanism (measure motor coil resistance) will be examined.

Commit:
Enable E-motor cool mode in farm mode only (and experimental menu)

Commit:
Enable E-motor current scaling mode for non-farm prints IFF ...

... the experimental menu is visible AND the EEPROM_ECOOL variable has a value of the universal answer to all problems of the universe - i.e. two conditions must be met at the start of the FW to enable the E-cool mode.

If the user enables the experimental menu, sets the E-cool mode and disables the menu afterwards, on the next start of the FW the E-cool mode will be DISABLED.

This is still subject to discussion how much obscure (security through obscurity) we'd like this option to have .

Commit:
Add serial debug msg to verify if E-cool mode is on

Commit:
Fixes after dev tests

DRracer reviewed (approved):
Approving changes done by @mkbel

mkbel commented on Firmware/ultralcd.cpp:
```diff
@@ -8947,6 +8954,14 @@ void lcd_experimental_toggle()
     eeprom_update_byte((uint8_t *)EEPROM_EXPERIMENTAL_VISIBILITY, oldVal);
 }
 
+void ECool_toggle(){
+    enableECool = ! enableECool;
+    // this is only called when the experimental menu is visible, thus the first condition for enabling of the ECool mode is met in this place
+    eeprom_update_byte((uint8_t *)EEPROM_ECOOL_ENABLE, enableECool ? EEPROM_ECOOL_MAGIC_NUMBER : EEPROM_EMPTY_VALUE);
+}
+bool ECool_get(){
+    return enableECool;
+}
```
It is overly complicated and error prone to cache EEPROM in RAM, and it also costs RAM. It is more efficient to just read status from EEPROM inside ECool_get() and use ECool_get() consistently and remove global variable enableECool.

Also at this point enableECool variable is sometimes used directly and sometimes through ECool_get().

If you remove enableECool variable, you can also remove its initialization from setup(), making things less error prone.

mkbel replied:
Also you can call TMC reinit from ECool_toggle(), so change is applied when the state of the variable changes. If experimental menu is not accessible during print, it is safe.

mkbel commented on Firmware/Marlin_main.cpp:
```diff
@@ -1337,13 +1337,15 @@ void setup()
 	tmc2130_mode = silentMode?TMC2130_MODE_SILENT:TMC2130_MODE_NORMAL;
 	update_mode_profile();
-	tmc2130_init();
+	tmc2130_init(TMCInitParams(false, farm_mode || enableECool));
```
It would be less error prone to hide (farm_mode || enableECool) condition inside ECool_get()

DRracer replied:
ok, good point

mkbel commented on Firmware/tmc2130.cpp:
```diff
@@ -191,10 +188,22 @@ void tmc2130_init()
-		tmc2130_setup_chopper(axis, tmc2130_mres[axis], tmc2130_current_h[axis], tmc2130_current_r[axis]);
+		tmc2130_setup_chopper(axis,
+            tmc2130_mres[axis], // this is not changed with ECool
+            tmc2130_current_h[axis] + params.enableECool,
+            tmc2130_current_r[axis] + params.enableECool);
```
Whiskey Tango Foxtrot?
Did you mean tmc2130_current_h[axis + params.enableECool] ?

DRracer replied:
OMG, yes I did

mkbel reviewed (commented):
M907 E gcode is not ignored when E motor scaling current is enabled.

mkbel commented on Firmware/tmc2130.h:
```diff
@@ -1,12 +1,14 @@
 extern uint8_t tmc2130_mode;
+// Beware - the 4th index (counted from zero) is abused for the E-motor cool mode only used on the farm
 //holding and running currents
-extern uint8_t tmc2130_current_h[4];
-extern uint8_t tmc2130_current_r[4];
+extern uint8_t tmc2130_current_h[5];
+extern uint8_t tmc2130_current_r[5];
```
This is not workable solution. If done this way, you would need to add condition to use index 3 or 4 on every place where this value is used, not only in tmc2130_init(). There are about 16 places in code where this value is used.

More workable solution would be to use approach already used in farm_mode (preferably use named constant instead of plain "36"):
```
		//increased extruder current (PFW363)
		tmc2130_current_h[E_AXIS] = 36;
		tmc2130_current_r[E_AXIS] = 36;
```

Commit:
Fix issues from code review

Commit:
Avoid access to E-cool mode switch on machines without TMC2130

mkbel commented on Firmware/variants/1_75mm_MK3S-EINSy10a-E3Dv6full.h:
```diff
@@ -267,7 +273,8 @@
 #define TMC2130_CURRENTS_H {16, 20, 35, 30}  // default holding currents for all axes
-#define TMC2130_CURRENTS_R {16, 20, 35, 30}  // default running currents for all axes
+#define TMC2130_CURRENTS_FARM 36             // E 805 mA peak for ECool/farm mode
+#define TMC2130_CURRENTS_R {16, 20, 35, 30}  // default running currents for all axes, the last parameter is E 805 mA peak for ECool mode
```
Comment after the comma `, the last parameter is E 805 mA peak for ECool mode`
is misleading now.

mkbel commented on Firmware/Marlin_main.cpp:
```diff
@@ -1337,13 +1340,14 @@ void setup()
 #endif //TMC2130_VARIABLE_RESOLUTION
 
 #endif //TMC2130
-
+    // We enable E-cool mode for non-farm prints IFF the experimental menu is visible AND the EEPROM_ECOOL variable has
+    // a value of the universal answer to all problems of the universe
```
Orphaned comment.

mkbel commented on Firmware/variants/1_75mm_MK3-EINSy10a-E3Dv6full.h:
```diff
@@ -247,6 +252,7 @@
 #define TMC2130_PWM_CLK   (2 * TMC2130_FCLK / TMC2130_PWM_DIV) // PWM frequency (23.4kHz, 35.1kHz, 46.9kHz, 58.5kHz for 12MHz fclk)
 
 #define TMC2130_TPWMTHRS  0         // TPWMTHRS - Sets the switching speed threshold based on TSTEP from stealthChop to spreadCycle mode
+#define TMC2130_TPWMTHRS_E 403
```
Suggest comment:
`// Switch extruder from StealthChop to SpreadCycle at around 900mm/min`

mkbel commented on Firmware/variants/1_75mm_MK3S-EINSy10a-E3Dv6full.h:
```diff
@@ -249,6 +254,7 @@
 #define TMC2130_PWM_CLK   (2 * TMC2130_FCLK / TMC2130_PWM_DIV) // PWM frequency (23.4kHz, 35.1kHz, 46.9kHz, 58.5kHz for 12MHz fclk)
 
 #define TMC2130_TPWMTHRS  0         // TPWMTHRS - Sets the switching speed threshold based on TSTEP from stealthChop to spreadCycle mode
+#define TMC2130_TPWMTHRS_E 403
```
Suggest comment:
`// Switch extruder from StealthChop to SpreadCycle at around 900mm/min`

mkbel commented on Firmware/ultralcd.cpp:
```diff
@@ -8947,6 +8947,35 @@ void lcd_experimental_toggle()
 }
 
+#ifdef TMC2130
+void UserECool_toggle(){
```
User? Did you mean Use?

DRracer replied:
nope, User - _like ordinary non-farm user-selected/activated_

mkbel commented on Firmware/ultralcd.cpp:
```diff
@@ -8947,6 +8947,35 @@ void lcd_experimental_toggle()
+/// change motors for various reasons and unless the motor is rotating, we cannot verify its properties
+/// (which would be obviously too late for an improperly sized motor)
+/// For farm printing, the cooler E-motor is enabled by default.
+bool UserECoolEnabled(){
```
User? Did you mean Use?

DRracer replied:
nope, User - like _ordinary non-farm user-selected/activated_

mkbel commented on Firmware/ultralcd.h:
```diff
@@ -133,6 +133,9 @@ extern uint8_t farm_mode;
 extern int farm_timer;
 extern uint8_t farm_status;
 
+extern bool UserECoolEnabled();
+extern bool FarmOrUserECool();
```
I think it is misleading to have 2 functions there, someone could use wrong one.
I suggest remove FarmOrUserECool() and hide its logic inside UserECoolEnabled(). I don't know what is "User" there. Why it is not called UseECool(), or ECoolEnabled()? Effect will be, that in farm_mode hidden menu will always show ECool [On] and toggling will left it [On] but it is absolutely ok for me.

DRracer replied:
I don't like the idea of merging the farm and user-selected (that's why I called it "User") E-cool mode, I'd like to keep it as it is.

Commit:
Fix comments from the last review

mkbel commented on Firmware/Marlin_main.cpp:
```diff
@@ -8440,17 +8443,19 @@ SERIAL_PROTOCOLPGM("\n\n");
-				uint8_t val = tmc2130_cur2val(cur_mA);
-				tmc2130_set_current_h(i, val);
-				tmc2130_set_current_r(i, val);
-				//if (i == E_AXIS) printf_P(PSTR("E-axis current=%ldmA\n"), cur_mA);
-			}
-
+        if( ! FarmOrUserECool() ){
+            // See tmc2130_cur2val() for translation to 0 .. 63 range
+            for (int i = 0; i < NUM_AXIS; i++){
+                if(code_seen(axis_codes[i]))
+                {
+                    long cur_mA = code_value_long();
+                    uint8_t val = tmc2130_cur2val(cur_mA);
+                    tmc2130_set_current_h(i, val);
+                    tmc2130_set_current_r(i, val);
+                    //if (i == E_AXIS) printf_P(PSTR("E-axis current=%ldmA\n"), cur_mA);
+                }
+            }
+        }
```
I think we want to disable only M907 E, not M907 XYZ.

Commit:
Do not allow only M907 E in case of E-cool mode

+warn the user on the serial line that the command was skipped

dafi87 commented:
With E Cool enabled, I feel like the extruder stepping got louder in stealth mode on my MK3S+. Was that just my ears or a bad print sample I picked, or did you guys notice similar behaviour?

Best,
Daniel

DRracer commented:
@dafi87 yes, the E-motor is "clicking" on fast changes of direction in this mode of operation... especially on retract/unretract moves.

dafi87 commented:
Thanks for getting back so quickly! Retract/unretract is exactly the situations where I thought I heard the difference in noise level.
